src/main/java/fi/aalto/cs/apluscourses/model/UrlRenderer.java

```diff
@@ -2,11 +2,13 @@
 
 import com.intellij.ide.BrowserUtil;
 import java.net.URI;
+import java.net.URISyntaxException;
+
 import org.jetbrains.annotations.NotNull;
 
 public class UrlRenderer {
 
-  public void show(@NotNull String url) throws Exception {
+  public void show(@NotNull String url) throws URISyntaxException {
```

Comment:
usually the most narrow checked `exception` is being thrown (@jaakkonarhi  check [this](https://stackoverflow.com/questions/25780619/why-do-we-follow-narrow-to-broad-exception-catching-mecahnism-in-java) as a starting point)

Reply:
Right, a thing to remember, though: an overriding method in a subclass cannot "broaden" the exception type in the `throws` declaration (the type of the exception is said to be _covariant_, i.e., for subtypes it can be more specific but not more generic).

Anyway, plain `throws Exception` should be generally avoided.